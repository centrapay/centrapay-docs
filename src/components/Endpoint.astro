---
import Eyebrow from './Eyebrow.astro';
import Row from './Row.astro';
import Column from './Column.astro';
import CodePanel from './CodePanel.astro';
import { Code } from 'astro:components';
import createEndpoint from '../utils/createEndpoint';

const theme = 'material-theme-darker';
const { path, filename } = Astro.props;

let endpoint;
if (filename) {
  const file = await import(`../content/api/_endpoints/${filename}.js`);
  endpoint = createEndpoint(file.default);
}
---

<Row>
  <Column>
    <Eyebrow method={endpoint.method} path={path}/>
    <slot/>
  </Column>

  { endpoint ?
    <Column sticky>
      <CodePanel title="Request" method={endpoint.method} path={endpoint.path}>
        <Code code={endpoint.requests.curl.code} lang={endpoint.requests.curl.lang} theme={theme}/>
      </CodePanel>

      { endpoint.response &&
        <CodePanel title="Response">
          <Code code={endpoint.response} lang="json" theme={theme}/>
        </CodePanel>
      }
    </Column>
    :
    <Column sticky>
      <slot name="code-examples"/>
    </Column>
  }
</Row>
