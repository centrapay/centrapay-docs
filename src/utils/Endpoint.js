import transformToHarRequest from './transformToHarRequest';
import { HTTPSnippet } from 'httpsnippet';

const supportedClients = {
  curl: {
    targetId: 'shell',
    options: { short: true, indent: ' ', prettifyJson: true },
  },
};

class Endpoint {
  constructor(props) {
    this.path = props.path;
    this.method = props.method;
    this.requests = props.requests;
  }

  static create(data) {
    const harRequest = transformToHarRequest(data);

    const snippet = new HTTPSnippet(harRequest);
    const path = snippet.requests[0].uriObj.pathname;
    const method = snippet.requests[0].method;

    const requests = {};
    Object.entries(supportedClients).forEach(([clientId, details]) => {
      requests[clientId] = {
        lang: details.targetId,
        code: snippet.convert(details.targetId, clientId, details.options)
      };
    });

    return new Endpoint({
      path,
      method,
      requests,
    });
  }
};

export default Endpoint;
