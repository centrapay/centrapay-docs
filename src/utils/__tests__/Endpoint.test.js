import Endpoint from '../Endpoint.js';

describe('Endpoint', () => {
  describe('createRequest', () => {
    it('creates a curl snippet', () => {
      const snippet = Endpoint.create({
        method: 'POST',
        path: '/api/test',
        request: {
          headers: {
            'X-Api-Key': '<TOKEN>',
            'Content-Type': 'application/json',
          },
          payload: {
            foo: 'bar',
            bar: 'baz',
          }
        }
      });
      expect(snippet.method).toBe('POST');
      expect(snippet.path).toBe('/api/test');
      expect(snippet.requests.curl).toMatchSnapshot();
    });
  });
});
