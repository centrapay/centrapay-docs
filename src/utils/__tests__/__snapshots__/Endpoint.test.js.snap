// Vitest Snapshot v1

exports[`Endpoint > createRequest > creates a curl snippet 1`] = `
{
  "code": "curl -X POST \\\\
 https://service.centrapay.com/api/test \\\\
 -H 'content-type: application/json' \\\\
 -H 'x-api-key: <TOKEN>' \\\\
 -d '
{
 \\"foo\\": \\"bar\\",
 \\"bar\\": \\"baz\\"
}
'",
  "lang": "shell",
}
`;
